pkg/httpclient/http.go

```diff
@@ -0,0 +1,130 @@
+// Copyright © 2021 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httpclient
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"runtime"
+	"time"
+
+	"github.com/gregjones/httpcache"
+	"go.thethings.network/lorawan-stack/v3/pkg/config/tlsconfig"
+	"go.thethings.network/lorawan-stack/v3/pkg/version"
+)
+
+// defaultHTTPClientTimeout is the default timeout for the HTTP client.
+const defaultHTTPClientTimeout = 10 * time.Second
+
+// TLSClientConfigurationProvider provides a *tls.Config to be used by TLS clients.
+type TLSClientConfigurationProvider interface {
+	GetTLSClientConfig(context.Context, ...tlsconfig.Option) (*tls.Config, error)
+}
+
+// Provider constructs *http.Clients.
+type Provider interface {
+	HTTPClient(context.Context, ...Option) (*http.Client, error)
+}
+
+// Option is an option for HTTP clients.
+type Option func(*httpClientOptions)
+
+type httpClientOptions struct {
+	cache            bool
+	tlsConfig        *tls.Config
+	tlsConfigOptions []tlsconfig.Option
+}
+
```

Comment:
Why not make the `httpClientOptions` exported? That way other packages could implement their own private funcs that could be passed in the instantiation of a new client. I understand that most operations functions that alter the http client would stay in the package but it does bring extensibility to it.

Don't know if it's necessary, but the following code example might give a better idea of what I'm saying.
https://gist.github.com/nicholaspcr/e647607ad7ca20c6bb14451ad62665a8

Reply:
Since we basically have a monoropo, if we need new options we just add them to the package itself - we don't really extend via 'inheritance'. Also keep in mind that the options that you provided can be applied even after the client has been built, which is not concurrency safe. This is the reason why they generally operate using an unexported function / struct.

Reply:
True, if exported there would be the potential of using the function outside the `httpclient` method, specially since it might not be intuitive to use. Then the current implementation looks good to me.